database/src/main/java/com/firebase/ui/database/FirebaseArrayOfObjects.java

```diff
@@ -0,0 +1,237 @@
+                                                            SnapshotParser<T> parser) {
+        FirebaseArrayOfObjects<T> array = newInstance(snapshots, modelClass);
+        array.mParser = parser;
+        return array;
+    }
+
+    public List<DataSnapshot> getSnapshots() {
+        return mSnapshots;
+    }
+
+    protected List<E> getObjects() {
+        List<E> objects = new ArrayList<>(mSnapshots.size());
+        for (int i = 0; i < mSnapshots.size(); i++) {
+            objects.add(get(i));
+        }
+        return objects;
+    }
+
+    @Override
+    public int size() {
+        return mSnapshots.size();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return mSnapshots.isEmpty();
+    }
+
+    @Override
+    public boolean contains(Object o) {
+        return indexOf(o) >= 0;
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @return an immutable iterator
+     */
+    @Override
+    public Iterator<E> iterator() {
+        return new ImmutableIterator(getObjects().iterator());
+    }
+
+    @Override
+    public Object[] toArray() {
+        return getObjects().toArray();
+    }
+
+    @Override
+    public <T> T[] toArray(T[] a) {
+        return getObjects().toArray(a);
+    }
+
+    @Override
+    public boolean containsAll(Collection<?> c) {
+        return getObjects().containsAll(c);
+    }
+
+    @Override
+    public E get(int index) {
+        return mParser.parseSnapshot(mSnapshots.get(index));
+    }
+
+    @Override
+    public int indexOf(Object o) {
+        return getObjects().indexOf(o);
+    }
+
+    @Override
+    public int lastIndexOf(Object o) {
+        return getObjects().lastIndexOf(o);
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @return an immutable list iterator
+     */
+    @Override
+    public ListIterator<E> listIterator() {
+        return new ImmutableListIterator(getObjects().listIterator());
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @return an immutable list iterator
+     */
+    @Override
+    public ListIterator<E> listIterator(int index) {
+        return new ImmutableListIterator(getObjects().listIterator(index));
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) return true;
+        if (obj == null || getClass() != obj.getClass()) return false;
+
+        FirebaseArrayOfObjects<?> array = (FirebaseArrayOfObjects<?>) obj;
+
+        return mSnapshots.equals(array.mSnapshots) && mEClass.equals(array.mEClass);
+    }
+
+    @Override
+    public int hashCode() {
+        int result = mSnapshots.hashCode();
+        result = 31 * result + mEClass.hashCode();
+        return result;
+    }
+
+    @Override
+    public String toString() {
+        return "FirebaseArrayOfObjects{" +
+                "mSnapshots=" + mSnapshots +
+                '}';
+    }
+
+    protected static class FirebaseArrayOfObjectsOptimized<E> extends FirebaseArrayOfObjects<E>
```

Comment:
@samtstern Totally agree with you about the comments, but I don't even know if it works yet so I'd rather wait so I don't cement possibly false ideas into my head. I've added it to the TODO list.

As for moving it to a separate class, we could do that but my thoughts were that it's supposed to be part of the same class and the same `newInstance` api. I didn't want to move it to a separate class because I don't want it to be public, but I also want people to be able to extend it, thus the `protected` modifier. What do you think?